elyra/util/operation_parser.py

```diff
@@ -111,53 +99,89 @@ def __init__(self, operation_filepath):
     def type(self):
         raise NotImplementedError()
 
+    @property
+    def model(self):
+        return self.build_model()
+
     @property
     def operation_filepath(self):
         # TODO : Ensure this is getting the abs path
         return self._operation_filepath
 
+    @property
+    @abstractmethod
+    def regex_dict(self) -> Dict[str, List]:
+        # a dictionary of matching regex expressions to use keyed to the json stanza
+        # Parsing for ENVs e.g. regex_dict['NAME_OF_JSON_STANZA'] = ['REGEX_EXP_1', 'REGEX_EXP_2']
+        raise NotImplementedError()
+
     @abstractmethod
-    def env_list(self):
+    def file_as_code_chunks(self) -> List[str]:
+        # A list of code `chunks` to look through. The size and length of the `chunks` is determined
+        # usually by the language. For now the concrete implementations below is by line-by-line in file
         raise NotImplementedError()
 
+    def build_model(self):
+        """
+        Builds a model dictionary of all the regex matches it for each key in the regex dictionary
+        :return:
+        """
+        model = dict()
+        for chunk in self.file_as_code_chunks():
+            for key in self.regex_dict.keys():
+                regex_list = self.regex_dict[key]
+                for pattern in regex_list:
+                    match_found = re.search(pattern, chunk)
+                    if match_found and match_found.group(1) not in model[key]:
+                        model[key].append(match_found.group(1))
+
+        return model
+
+
+class NotebookOperationParser(OperationParser):
+
+    _type = "notebook"
+
+    def __init__(self, operation_filepath):
+        super().__init__(operation_filepath)
+        self.notebook = nbformat.read(open(self.operation_filepath), as_version=4)
+        self.language = self.notebook['metadata']['kernelspec']['language']
+
+    def type(self):
+        return self._type
+
+    def file_as_code_chunks(self) -> List[str]:
+        for cell in self.notebook.cells:
+            if cell.source:
+                file_by_line = cell.source.split('\n')
+                return file_by_line
+
+    def regex_dict(self) -> Dict[str, List]:
+        if self.language == 'python':
+            return PythonOperationParser(self.operation_filepath).regex_dict()
```

Comment:
I think it might be good to also have a more granular language-based parser.  For example, the "operation parsers" are responsible for feeding the language parses content to parse.

For example, the PythonScriptOperationParser would then feed a line (or list of lines) to a PythonEnvParser.  The xxxEnvParser then knows how to extract env variables from a list of lines for its language.  If that extraction requires regular expression matches, then so be it - that's an implementation detail of the language parser.

The NotebookOperationParser would also _contain_ an instance of a language-based parser depending on the language of the notebook and feed it a list of lines pertaining to a given code cell.